lxd/images.go

```diff
@@ -2993,37 +2992,65 @@ func imageGet(d *Daemon, r *http.Request) response.Response {
 		return response.SmartError(err)
 	}
 
-	// Get the image (expand partial fingerprints).
+	trusted := auth.IsTrusted(r.Context())
+	secret := r.FormValue("secret")
+
+	// Unauthenticated clients that do not provide a secret may only view public images.
+	publicOnly := !trusted && secret == ""
+
+	// Get the image. We need to do this before the permission check because the URL in the permission check will not
+	// work with partial fingerprints.
 	var info *api.Image
 	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
-		info, err = doImageGet(ctx, tx, projectName, fingerprint, false)
+		info, err = doImageGet(ctx, tx, projectName, fingerprint, publicOnly)
 		if err != nil {
 			return err
 		}
 
 		return nil
 	})
-	if err != nil {
+	if err != nil && api.StatusErrorCheck(err, http.StatusNotFound) {
+		// Return a generic not found. This is so that the caller cannot determine the existence of an image by the
```

Comment:
What is the non-generic error then?

Reply:
The non-generic error would be the `response.SmartError` in the following `else if err != nil` block. It could contain any error that occurred during the transaction.

Reply:
What made you hide this into a generic one?

Reply:
The authorizer returns a generic not found error if the caller doesn't have permission to view a resource. In this case, because we have to expand the fingerprint before we can call the authorizer, we don't want to leak any errors indicating the existence of the image. I'm just trying to be defensive with it.

Reply:
got you thanks

Reply:
I'm not sure I've explained this clearly. 

Current behaviour (trusted without permission or untrusted):
- `GET /1.0/images/{non-existing-image}` = `Image not found`
- `GET /1.0/images/{existing-image}` = `Not found`

Both cases return a 404, but you can determine that the image does in fact exist because the error message when it really doesn't exist is different from the error message when you're not allowed to view it.

Reply:
This is the explanation i was hoping for, thanks!